### PR TITLE
Compiler can now emit warnings

### PR DESCRIPTION
**Pull request by tomwhatmore**
Addresses #479

### CODE REVIEW

**Comment by lpil** [comment]
This looks good! Thanks!

I want the warnings to be emitted whether the module succeeds compilation or not, and I don't think we should cache the warnings in #157 , so I don't think we should be storing them on the module.

For this PR we could have the module typing function return `(Result<a, b>, Vec<Warning>)`, though later I think we could either create some kind of telemetry object and pass it in, or we can have it as a property on Env (which I would like to restructure to be an exposed object called Typer, similar to the Formatter and Printer)

**Comment by tomwhatmore** [comment]
Took the warnings out of the modules and now returning them seperately from `infer_module()`

**Comment by lpil** [comment]
Build is failing FYI

**Comment by tomwhatmore** [comment]
Whoops! Not sure what happened there, I'll fix it when I've made some
dinner.

**Comment by tomwhatmore** [comment]
Apparently I deleted the warnings field from the generated `grammar.rs` file instead of `grammar.lalrpop`. Blunder.

**Comment by lpil** [comment]
A classic move! I have to work out how to get my editor from automatically opening that file when it has an error in it.

File: src/project.rs [hunk 1]
```diff
@@ -92,8 +95,21 @@ pub fn analysed(inputs: Vec<Input>) -> Result<Vec<Analysed>, Error> {
 
         println!("Compiling {}", name_string);
 
-        let ast = crate::typ::infer_module(module, &modules_type_infos)
-            .map_err(|error| Error::Type { path, src, error })?;
+        let mut warnings = Vec::new();
+        let (result, mut module_warnings) = crate::typ::infer_module(module, &modules_type_infos);
+        module_warnings.drain(..).for_each(|warning| {
```
Comment (lpil) [inline src/project.rs:100]: Could this be `module_warnings.map` ? Would be slightly more concise and the vector wouldn't be grown as new warnings are added.

Comment (tomwhatmore) [reply src/project.rs:100]: Yeah, probably. `.map` was being a pain when it was part of the module struct but might be ok now. Will give it a go.

Comment (lpil) [reply src/project.rs]: Thanks

Comment (tomwhatmore) [reply src/project.rs]: Yeah that works fine, I was just being dumb earlier. Updated.

Comment (lpil) [reply src/project.rs]: I'm sure you wern't :)

**Pull request merged**